enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -1348,6 +1365,105 @@ class AdjointGenerator
       addToDiffe(orig_op1, dif1, Builder2, addingType);
   }
 
+  void createBinaryOperatorDual(llvm::BinaryOperator &BO) {
+    IRBuilder<> Builder2(&BO);
+
+    Instruction *nBO = gutils->getNewFromOriginal(&BO);
+
+    assert(nBO);
+    assert(nBO->getNextNode());
+
+    if (nBO->getNextNode()) {
+      Builder2.SetInsertPoint(nBO->getNextNode());
+    }
+
+    Builder2.SetCurrentDebugLocation(
+        gutils->getNewFromOriginal(Builder2.getCurrentDebugLocation()));
+    Builder2.setFastMathFlags(getFast());
+
+    Value *orig_op0 = BO.getOperand(0);
+    Value *orig_op1 = BO.getOperand(1);
+
+    bool constantval0 = gutils->isConstantValue(orig_op0);
+    bool constantval1 = gutils->isConstantValue(orig_op1);
+
+    Value *dif0 = constantval0 ? nullptr : diffe(orig_op0, Builder2);
+    Value *dif1 = constantval1 ? nullptr : diffe(orig_op1, Builder2);
+
+    Type *addingType = BO.getType();
+
+    switch (BO.getOpcode()) {
+    case Instruction::FMul: {
+      if (constantval0) {
+        Value *idiff = Builder2.CreateFMul(orig_op0, dif0);
+        setDiffe(&BO, idiff, Builder2);
+      } else if (constantval1) {
+        Value *idiff = Builder2.CreateFMul(orig_op1, dif1);
+        setDiffe(&BO, idiff, Builder2);
+      } else {
+        Value *idiff0 =
+            Builder2.CreateFMul(dif0, gutils->getNewFromOriginal(orig_op0));
+        Value *idiff1 =
+            Builder2.CreateFMul(dif1, gutils->getNewFromOriginal(orig_op1));
+        addToDiffe(&BO, idiff0, Builder2, addingType);
+        addToDiffe(&BO, idiff1, Builder2, addingType);
+      }
+      break;
+    }
+    case Instruction::FAdd: {
+      if (!constantval0) {
+        addToDiffe(&BO, dif0, Builder2, addingType);
+      }
+
+      if (!constantval1) {
+        addToDiffe(&BO, dif1, Builder2, addingType);
+      }
+      break;
+    }
+    case Instruction::FSub: {
+      if (!constantval0) {
+        addToDiffe(&BO, dif0, Builder2, addingType);
+      }
+
+      if (!constantval1) {
+        addToDiffe(&BO, Builder2.CreateFNeg(dif1), Builder2, addingType);
+      }
+      break;
+    }
+    case Instruction::FDiv: {
+      Value *idiff1;
+      if (!constantval0) {
+        idiff1 =
+            Builder2.CreateFMul(dif0, gutils->getNewFromOriginal(orig_op1));
+      } else {
+        idiff1 = ConstantFP::get(addingType, 0.0);
+      }
+
+      Value *idiff2;
+      if (!constantval1) {
+        idiff2 =
+            Builder2.CreateFMul(gutils->getNewFromOriginal(orig_op0), dif1);
+      } else {
+        idiff2 = ConstantFP::get(addingType, 0.0);
```

Comment:
If something is inactive (constant) then there isn't a shadow for you to update, right?